### ISSUE TITLE
Bytes may recreate Vec incorrectly on drop

### ISSUE THREAD

**Issue by stepancheg** [tokio-rs/bytes#360]
This example:

```
diff --git a/examples/rrr.rs b/examples/rrr.rs
index e69de29..60374de 100644
--- a/examples/rrr.rs
+++ b/examples/rrr.rs
@@ -0,0 +1,7 @@
+use bytes::Bytes;
+
+fn main() {
+    let mut bytes = Bytes::from(vec![10, 20, 30]);
+    bytes.truncate(2);
+    drop(bytes);
+}
diff --git a/src/bytes.rs b/src/bytes.rs
index 93ab84b..6af5148 100644
--- a/src/bytes.rs
+++ b/src/bytes.rs
@@ -767,6 +767,7 @@ impl From<Vec<u8>> for Bytes {
         let slice = vec.into_boxed_slice();
         let len = slice.len();
         let ptr = slice.as_ptr();
+        println!("dismantling a vec of cap {}", len);
         drop(Box::into_raw(slice));
 
         if ptr as usize & 0x1 == 0 {
@@ -886,6 +887,7 @@ unsafe fn promotable_odd_drop(data: &mut AtomicPtr<()>, ptr: *const u8, len: usi
 
 unsafe fn rebuild_vec(buf: *mut u8, offset: *const u8, len: usize) -> Vec<u8> {
     let cap = (offset as usize - buf as usize) + len;
+    println!("rebuilding vec with cap {}", cap);
     Vec::from_raw_parts(buf, cap, cap)
 }
 
diff --git a/src/lib.rs b/src/lib.rs
index 2df1fb3..5d9355b 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,6 +1,5 @@
 #![deny(warnings, missing_docs, missing_debug_implementations, rust_2018_idioms)]
 #![doc(html_root_url = "https://docs.rs/bytes/0.5.3")]
-#![no_std]
 
 //! Provides abstractions for working with bytes.
 //!
```

```
% cargo run --example rrr
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/examples/rrr`
dismantled a vec of cap 3
rebuilding vec with cap 2
```

So we are allocating 3 * 4 bytes, but when calling deallocation we are passing hint parameter 2 * 4 which causes deallocation by `RawVec` which calls `GlobalAlloc.dealloc` which violates the spec: ("layout must be the same layout that was used to allocate that block of memory").

I suspect it may crash deallocation or cause memory leaks, although I could not reproduce either.

I tried this program:

```
#[global_allocator]
static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;

use bytes::Bytes;

fn main() {
    loop {
        let mut bytes = Bytes::from(vec![10; 100000000]);
        bytes.truncate(1);
        drop(bytes);
    }
}
```

After a few seconds of run, macOS Activity Monitor shows 100Gb of memory usage by this program which is a sign that something doesn't work as expected.

**Comment by udoprog** [comment]
Can verify that the layout is indeed incorrect, we need to use the capacity of the vector (rather than the truncated length).

This causes issues specifically with `Jemalloc` because it uses APIs which rely on user-provided metadata as an optimization (`mallocx`, `sdallocx`). Likely so it doesn't have to maintain it's own internal structures. See:
https://docs.rs/jemallocator/0.3.2/src/jemallocator/lib.rs.html#77

**Comment by seanmonstar** [comment]
Yep, I was able to add a unit test for the Vec repr's allocation usage. There's a proposed fix in #361.
